### PR TITLE
Implement `dpnp.interp()`

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR suggests adding an implementation of `dpnp.interp()` with sycl_kernel including support for all function arguments.

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] Have you added documentation for your changes, if necessary?
- [ ] Have you added your changes to the changelog?

### CODE REVIEW

File: dpnp/backend/kernels/elementwise_functions/interpolate.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                             const TValue *left,
+                             const TValue *right,
+                             TValue *out,
+                             const std::size_t n,
+                             const std::size_t xp_size,
+                             const std::vector<sycl::event> &depends)
+{
+    return q.submit([&](sycl::handler &h) {
+        h.depends_on(depends);
+        h.parallel_for(sycl::range<1>(n), [=](sycl::id<1> i) {
```
Comment (antonwolfy) [inline dpnp/backend/kernels/elementwise_functions/interpolate.hpp:30]: It would be helpful to leave a comment why it was decided to use `sycl::id<1>`, but not `sycl::nd_item<1>` as we used in many other kernels.

File: dpnp/tests/test_sycl_queue.py [hunk 2]
```diff
@@ -1453,6 +1453,24 @@ def test_choose(device):
     assert_sycl_queue_equal(result.sycl_queue, chc.sycl_queue)
 
 
+@pytest.mark.parametrize("device", valid_dev, ids=dev_ids)
+@pytest.mark.parametrize("left", [None, dpnp.array(-1.0)])
+@pytest.mark.parametrize("right", [None, dpnp.array(99.0)])
+@pytest.mark.parametrize("period", [None, dpnp.array(180.0)])
```
Comment (antonwolfy) [inline dpnp/tests/test_sycl_queue.py:1459]: no need to have that as array here, because there will be `dpnp.array(..., sycl_queue=x.sycl_queue)` anyway
```suggestion
@pytest.mark.parametrize("left", [None, -1.0])
@pytest.mark.parametrize("right", [None, 99.0])
@pytest.mark.parametrize("period", [None, 180.0])
```

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+                   const dpctl::tensor::usm_ndarray &idx,
+                   const dpctl::tensor::usm_ndarray &xp,
+                   const dpctl::tensor::usm_ndarray &fp,
+                   std::optional<const dpctl::tensor::usm_ndarray> &left,
+                   std::optional<const dpctl::tensor::usm_ndarray> &right,
+                   dpctl::tensor::usm_ndarray &out,
+                   sycl::queue &exec_q,
+                   const std::vector<sycl::event> &depends)
+{
+    if (x.get_size() == 0) {
```
[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:196]: based on python code:
```python
    if xp.size == 0:
        raise ValueError("array of sample points is empty")
```
it seems as unsupported use case

File: dpnp/dpnp_iface_mathematical.py [hunk 4]
```diff
@@ -2742,6 +2777,180 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+        raise TypeError(
+            "Cannot cast array data from"
+            f" {x_dtype} to {x_float_type} "
+            "according to the rule 'safe'"
+        )
+
+    x = dpnp.asarray(x, dtype=x_float_type, order="C")
+    xp = dpnp.asarray(xp, dtype=x_float_type, order="C")
+
+    out_dtype = dpnp.common_type(x, xp, fp)
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2899]: I guess the result has at least default real floating data type, isn't it?
```suggestion
    out_dtype = dpnp.common_type(x_float_type, fp)
```

File: dpnp/dpnp_iface_mathematical.py [hunk 5]
```diff
@@ -348,6 +349,40 @@ def _process_ediff1d_args(arg, arg_name, ary_dtype, ary_sycl_queue, usm_type):
...
+
+    if dpnp.is_supported_array_type(param):
+        if param.ndim != 0:
+            raise ValueError(
+                f"a {name} value must be 0-dimensional, "
+                f"but got {param.ndim}-dim"
+            )
+        if dpu.get_execution_queue([exec_q, param.sycl_queue]) is None:
+            raise ValueError(
+                "input arrays and {name} must be on the same SYCL queue"
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:369]:
```suggestion
                "input arrays and {name} must be allocated on the same SYCL queue"
```

File: dpnp/dpnp_iface_mathematical.py [hunk 6]
```diff
@@ -2742,6 +2777,180 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+    x = dpnp.asarray(x, dtype=x_float_type, order="C")
+    xp = dpnp.asarray(xp, dtype=x_float_type, order="C")
+
+    out_dtype = dpnp.common_type(x, xp, fp)
+
+    fp = dpnp.asarray(fp, dtype=out_dtype, order="C")
+
+    if period is not None:
+        period = _validate_interp_param(period, "period", exec_q, usm_type)
+        if period == 0:
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2905]: That check leads to SYCL queue synchronization. Is that expected?

File: dpnp/dpnp_iface_mathematical.py [hunk 7]
```diff
@@ -2742,6 +2777,180 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+        # normalizing periodic boundaries
+        x %= period
+        xp %= period
+        asort_xp = dpnp.argsort(xp)
+        xp = xp[asort_xp]
+        fp = fp[asort_xp]
+        xp = dpnp.concatenate((xp[-1:] - period, xp, xp[0:1] + period))
+        fp = dpnp.concatenate((fp[-1:], fp, fp[0:1]))
+        assert xp.flags.c_contiguous
+        assert fp.flags.c_contiguous
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2922]: What is the intention of that assertions?

File: dpnp/dpnp_iface_mathematical.py (hunk 4 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_mathematical.py:2899]: `dpnp.common_type` takes *arrays as input argument and I can not pass `x_float_type`

File: dpnp/dpnp_iface_mathematical.py (hunk 7 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_mathematical.py:2922]: To make sure that `xp` and `fp` are contiguous after normalizing periodic boundaries

Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2923]: But isn't `dpnp.concatenate` return a new array always? We never used the assertation to ensure contiguous of the data before passing to the extension and I wonder why it's needed here.

File: dpnp/backend/kernels/elementwise_functions/interpolate.hpp (hunk 1 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/kernels/elementwise_functions/interpolate.hpp:52]: This version of kernel implementation is selected over the work-group version because it shows slightly better performance likely due to simpler execution without additional group and loop overhead

Comparison of two versions with different WorkPI on PVC

image

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    array_names names = {{&x, "x"}, {&xp, "xp"}, {&fp, "fp"}, {&out, "out"}};
+
+    auto array_types = td_ns::usm_ndarray_types();
+    int x_type_id = array_types.typenum_to_lookup_id(x.get_typenum());
+    int idx_type_id = array_types.typenum_to_lookup_id(idx.get_typenum());
+    int xp_type_id = array_types.typenum_to_lookup_id(xp.get_typenum());
+    int fp_type_id = array_types.typenum_to_lookup_id(fp.get_typenum());
+    int out_type_id = array_types.typenum_to_lookup_id(out.get_typenum());
+
+    if (x_type_id != xp_type_id) {
```
Comment (AlexanderKalistratov) [inline dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:123]: I'd suggest to add additional functions to `validation_utils`:
```C++
void check_same_dtype(const array_ptr &arr1,
        const array_ptr &arr2,
        const array_names &names)
{
    if (arr1 == nullptr || arr2 == nullptr) {
        return;
    }
    auto array_types = td_ns::usm_ndarray_types();

    auto first_type_id = array_types.typenum_to_lookup_id(arr1->get_typenum());
    auto second_type_id = array_types.typenum_to_lookup_id(arr2->get_typenum());

    if (first_type_id != second_type_id) {
        py::dtype first_dtype = dtype_from_typenum(first_type_id);
        py::dtype second_dtype = dtype_from_typenum(second_type_id);

        auto first_dtype_str = py::str(first_dtype);
        auto second_dtype_str = py::str(second_dtype);

        throw py::value_error(py::str("Arrays " + name_of(arr1, names) + " and " +
                              name_of(arr2, names) + " must have the same "
                              "dtype, but got ") +
                              first_dtype_str + py::str(" and ") +
                              second_dtype_str);
    }
}

void check_same_dtype(const std::vector<array_ptr> &arrays,
    const array_names &names)
{
    if (arrays.empty()) {
        return;
    }

    auto first = arrays[0];

    for (size_t i = 1; i < arrays.size(); ++i) {
        auto second = arrays[i];
        check_same_dtype(first, second, names);
    }
}
```

And call it here

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        throw py::value_error("x and xp must have the same dtype");
+    }
+    if (fp_type_id != out_type_id) {
+        throw py::value_error("fp and out must have the same dtype");
+    }
+    if (idx_type_id != static_cast<int>(td_ns::typenum_t::INT64)) {
```
Comment (AlexanderKalistratov) [inline dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:129]: You probably could add function to `validation utils`. Something like
```C++
void check_has_dtype(const array_ptr &arr,
    const typenum_t dtype,
    const array_names &names);
```

And call it here

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        throw py::value_error("The type of idx must be int64");
+    }
+
+    auto left_v = left ? &left.value() : nullptr;
+    if (left_v) {
+        names.insert({left_v, "left"});
+        if (left_v->get_ndim() != 0) {
```
Comment (AlexanderKalistratov) [inline dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:136]: You can do instead:
```C++
check_num_dims(left_v, 0, names);
```

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            array_types.typenum_to_lookup_id(right_v->get_typenum());
+        if (right_type_id != fp_type_id) {
+            throw py::value_error(
+                "right must have the same dtype as fp and out");
+        }
+    }
+
+    common_checks({&x, &xp, &fp, left_v, right_v}, {&out}, names);
+
+    if (x.get_ndim() != 1 || xp.get_ndim() != 1 || fp.get_ndim() != 1 ||
```
Comment (AlexanderKalistratov) [inline dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:165]: There is a single array version for `check_num_dims` but you could add vector version of it:
```C++
inline void check_num_dims(const std::vector<array_ptr> &arrays,
                           const size_t ndim,
                           const array_names &names)
{
    for (const auto &arr : arrays) {
        check_num_dims(arr, ndim, names);
    }
}
```

And then call it here:
```
check_num_dims({x, xp, fp, idx, out}, 1, names);
```

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        idx.get_ndim() != 1 || out.get_ndim() != 1)
+    {
+        throw py::value_error("All arrays must be one-dimensional");
+    }
+
+    if (xp.get_size() != fp.get_size()) {
```
Comment (AlexanderKalistratov) [inline dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:171]: Add function `check_same_size` and/or `check_same_shape`?

File: dpnp/backend/extensions/common/ext/common.hpp [hunk 13]
```diff
@@ -106,6 +106,18 @@ struct IsNan
...
+struct value_type_of
+{
+    using type = T;
+};
+
+template <typename T>
+struct value_type_of<std::complex<T>>
+{
+    using type = T;
+};
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/common/ext/common.hpp:119]: That is good point but I  would prefer `using type = typename std::remove_cv_t<T>::value_type`  to ensure it works correctly with `const` types

[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/backend/extensions/common/ext/common.hpp:119]: Do you mean something like?
```cpp
struct value_type_of<std::complex<T>>
{
    using type = typename std::remove_cv_t<T>::value_type;
};
```
Then, I believe, it will not work.

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp (hunk 10 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp]: done, thank you

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp (hunk 9 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp]: done, thank you

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp (hunk 11 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp]: done, thank you

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp (hunk 12 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp]: done, thank you

File: dpnp/backend/extensions/common/ext/common.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/common/ext/common.hpp:122]: I actually meant removing `const` in `value_type_of_impl<T, true>`

```cpp
template<typename T>
struct value_type_of_impl<T, true>
{
    using type = typename std::remove_cv_t<T>::value_type;
};
```

[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/common/ext/common.hpp:122]: Done in https://github.com/IntelPython/dpnp/pull/2417/commits/9e06cc35968823ce9fc5eecfd63f3b146c204ea6

File: dpnp/dpnp_iface_mathematical.py (hunk 6 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_mathematical.py:2907]: That is good catch!
I have updated this logic to avoid synchronization
Now `period` argument supports only `scalar` and `None`

File: dpnp/dpnp_iface_mathematical.py [hunk 14]
```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+    period : {None, scalar}, optional
+        A period for the x-coordinates. This parameter allows the proper
+        interpolation of angular x-coordinates. Parameters `left` and `right`
+        are ignored if `period` is specified.
+
+        Default: ``None``.
+
+    Returns
+    -------
+    y : {dpnp.ndarray, usm_ndarray}
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2825]: Do we really return `usm_ndarray`?

File: dpnp/dpnp_iface_mathematical.py [hunk 15]
```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+    x = dpnp.asarray(x, dtype=x_float_type, order="C")
+    xp = dpnp.asarray(xp, dtype=x_float_type, order="C")
+
+    out_dtype = dpnp.common_type(x, xp, fp)
+
+    fp = dpnp.asarray(fp, dtype=out_dtype, order="C")
+
+    if period is not None:
+        if not dpnp.isscalar(period):
+            raise TypeError(f"period must be a scalar, but got {type(period)}")
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2906]:
```suggestion
            raise TypeError(f"period must be a scalar or None, but got {type(period)}")
```

File: dpnp/dpnp_iface_mathematical.py [hunk 16]
```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+        if period == 0:
+            raise ValueError("period must be a non-zero value")
+        period = _validate_interp_param(period, "period", exec_q, usm_type)
+
+        period = dpnp.abs(period)
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2911]:
```suggestion
        period = _validate_interp_param(builtins.abs(period), "period", exec_q, usm_type)
```

File: dpnp/dpnp_iface_mathematical.py [hunk 17]
```diff
@@ -348,6 +349,41 @@ def _process_ediff1d_args(arg, arg_name, ary_dtype, ary_sycl_queue, usm_type):
     return arg, usm_type
 
 
+def _validate_interp_param(param, name, exec_q, usm_type, dtype=None):
+    """
+    Validate and convert optional parameters for interpolation.
+
+    Returns a USM array or None if the input is None.
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:356]: What is the intention to return usm_ndarray instead of dpnp.ndarray?

File: dpnp/dpnp_iface_mathematical.py [hunk 18]
```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+
+    if not dpnp.can_cast(x_dtype, x_float_type):
+        raise TypeError(
+            "Cannot cast array data from"
+            f" {x_dtype} to {x_float_type} "
+            "according to the rule 'safe'"
+        )
+
+    x = dpnp.asarray(x, dtype=x_float_type, order="C")
+    xp = dpnp.asarray(xp, dtype=x_float_type, order="C")
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2898]: dtype check is implemented as part of `astype`, so:
```suggestion
    x_float_type = dpnp.default_float_type(exec_q)
    x = dpnp.astype(x, x_float_type, order="C", casting="safe", copy=False)
    xp = dpnp.astype(xp, x_float_type, order="C", casting="safe", copy=False)
```

File: dpnp/dpnp_iface_mathematical.py [hunk 19]
```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+    One-dimensional linear interpolation.
+
+    Returns the one-dimensional piecewise linear interpolant to a function
+    with given discrete data points (`xp`, `fp`), evaluated at `x`.
+
+    For full documentation refer to :obj:`numpy.interp`.
+
+    Parameters
+    ----------
+    x : {dpnp.ndarray, usm_ndarray}
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:2792]: Do we need to state that both `x` and `xp` are expected to have a real-valued floating-point data type (see other ufuncs for a reference)?

File: dpnp/tests/test_mathematical.py [hunk 20]
```diff
@@ -1143,6 +1143,152 @@ def test_complex(self, xp):
...
+        xp = dpnp.array([0, 1])
+        fp = dpnp.array([[1, 2]])
+        assert_raises(ValueError, dpnp.interp, x, xp, fp)
+
+        # xp and fp are empty
+        xp = dpnp.array([])
+        fp = dpnp.array([])
+        assert_raises(ValueError, dpnp.interp, x, xp, fp)
+
+        # x complex
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1261]: Is there any test that `xp` can't be complex?

File: dpnp/tests/test_mathematical.py [hunk 21]
```diff
@@ -1143,6 +1143,152 @@ def test_complex(self, xp):
         assert_raises((ValueError, TypeError), xp.i0, a)
 
 
+class TestInterp:
+    @pytest.mark.parametrize(
+        "dtype_x", get_all_dtypes(no_bool=True, no_complex=True)
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1148]: Do we have tests with bool inputs?

File: dpnp/tests/test_mathematical.py [hunk 22]
```diff
@@ -1143,6 +1143,152 @@ def test_complex(self, xp):
...
+    )
+    @pytest.mark.parametrize("dtype_y", get_all_dtypes(no_bool=True))
+    def test_all_dtypes(self, dtype_x, dtype_y):
+        x = numpy.linspace(0.1, 9.9, 20).astype(dtype_x)
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1152]: Do we have any test where `x` and `xp` have different dtypes?

File: dpnp/tests/test_mathematical.py (hunk 21 above)
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1148]:
```suggestion
        "dtype_x", get_all_dtypes(no_bool=True, no_complex=True, no_none=True)
```

Comment (antonwolfy) [reply dpnp/tests/test_mathematical.py:1148]: No need to test `None` dtype here and after.

File: dpnp/tests/test_mathematical.py [hunk 23]
```diff
@@ -1143,6 +1143,152 @@ def test_complex(self, xp):
         assert_raises((ValueError, TypeError), xp.i0, a)
 
 
+class TestInterp:
+    @pytest.mark.parametrize(
+        "dtype_x", get_all_dtypes(no_bool=True, no_complex=True)
+    )
+    @pytest.mark.parametrize("dtype_y", get_all_dtypes(no_bool=True))
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1150]: Do we have a test with bool `fp` array?

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 24]
```diff
@@ -0,0 +1,263 @@
...
+                   std::optional<const dpctl::tensor::usm_ndarray> &left,
+                   std::optional<const dpctl::tensor::usm_ndarray> &right,
+                   dpctl::tensor::usm_ndarray &out,
+                   sycl::queue &exec_q,
+                   const std::vector<sycl::event> &depends)
+{
+    common_interpolate_checks(x, idx, xp, fp, out, left, right);
+
+    if (x.get_size() == 0) {
+        return {sycl::event(), sycl::event()};
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:165]: I will remove it.
Thanks

File: dpnp/tests/test_mathematical.py (hunk 23 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/test_mathematical.py]: Done

File: dpnp/dpnp_iface_mathematical.py (hunk 19 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_mathematical.py:2793]: done

File: dpnp/tests/test_mathematical.py (hunk 20 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/test_mathematical.py:1264]: done

File: dpnp/tests/test_mathematical.py (hunk 22 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/test_mathematical.py:1155]: done

File: dpnp/tests/test_mathematical.py [hunk 25]
```diff
@@ -1143,6 +1143,159 @@ def test_complex(self, xp):
...
 
 
+class TestInterp:
+    @pytest.mark.parametrize(
+        "dtype_x", get_all_dtypes(no_complex=True, no_none=True)
+    )
+    @pytest.mark.parametrize(
+        "dtype_xp", get_all_dtypes(no_complex=True, no_none=True)
+    )
+    @pytest.mark.parametrize("dtype_y", get_all_dtypes())
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1153]:
```suggestion
    @pytest.mark.parametrize("dtype_y", get_all_dtypes(no_none=True))
```

File: dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+ * @brief A factory to define pairs of supported types for which
+ * interpolate function is available.
+ *
+ * @tparam T Type of input vector `a` and of result vector `y`.
+ */
+template <typename T>
+struct InterpolateOutputType
+{
+    using value_type = typename std::disjunction<
+        td_ns::TypeMapResultEntry<T, sycl::half>,
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/backend/extensions/ufunc/elementwise_functions/interpolate.cpp:210]: You are right! It would be better to just delete it

File: dpnp/dpnp_iface_mathematical.py [hunk 27]
```diff
@@ -348,6 +350,41 @@ def _process_ediff1d_args(arg, arg_name, ary_dtype, ary_sycl_queue, usm_type):
     return arg, usm_type
 
 
+def _validate_interp_param(param, name, exec_q, usm_type, dtype=None):
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_mathematical.py:353]: done

File: dpnp/tests/test_mathematical.py [hunk 28]
```diff
@@ -1197,6 +1197,23 @@ def test_left_right_args(self, dtype):
         result = dpnp.interp(ix, ixp, ifp, left=-40, right=40)
         assert_dtype_allclose(result, expected)
 
+        # left and right as ndarray
```
Comment (antonwolfy) [inline dpnp/tests/test_mathematical.py:1200]: It can be done through `parametrize("left, right", [[-40, 40], [dpnp.array(-40), dpnp.array(40)]])` and then:
```python
expected = numpy.interp(x, xp, fp, left=get_array(numpy, left), right=get_array(numpy, right))
result = dpnp.interp(ix, ixp, ifp, left=left, right=right)
```
or something similar

File: dpnp/dpnp_iface_mathematical.py [hunk 29]
```diff
@@ -348,6 +350,41 @@ def _process_ediff1d_args(arg, arg_name, ary_dtype, ary_sycl_queue, usm_type):
...
+                f"but got {param.ndim}-dim"
+            )
+        if dpu.get_execution_queue([exec_q, param.sycl_queue]) is None:
+            raise ValueError(
+                f"input arrays and {name} must be allocated "
+                "on the same SYCL queue"
+            )
+        param = param.astype(dtype)
+
+        return param
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_mathematical.py:375]:
```suggestion
        return param.astype(dtype)
```

**Review by antonwolfy — APPROVED** [review]
Thank you @vlad-perevezentsev, LGTM with only a minor comment from me.
